src/stm.ts

```diff
@@ -0,0 +1,515 @@
+import BigNumber from 'bignumber.js';
+
+import * as Builder from './builder';
+import { ClientRuntimeError, STMConflictError } from './errors';
+import { Range } from './range';
+import * as RPC from './rpc';
+import { NSApplicator, toBuffer } from './util';
+
+/**
+ * Isolation level which can be passed into the ISTMOptions.
+ */
+export const enum Isolation {
+  /**
+   * SerializableSnapshot provides serializable isolation and
+   * also checks for write conflicts.
+   */
+  SerializableSnapshot,
+
+  /**
+   * Serializable reads within the same transaction attempt return data
+   * from the at the revision of the first read.
+   */
+  Serializable,
+
+  /**
+   * RepeatableReads reads within the same transaction attempt always
+   * return the same data.
+   */
+  RepeatableReads,
+
+  /**
+   * ReadCommitted reads keys from any committed revision.
+   */
+  ReadCommitted,
+}
+/**
+ * ISTMOptions are optionally passed to `etcd3.stm(options)`.
+ */
+export interface ISTMOptions {
+  /**
+   * Number of times we'll retry the transaction if we get a conflict.
+   * Defaults to 3.
+   */
+  retries: number;
+
+  /**
+   * WithPrefetch is a hint to prefetch a list of keys before trying to apply.
+   * If an STM transaction will unconditionally fetch a set of keys, prefetching
+   * those keys will save the round-trip cost from requesting
+   * each key one by one with `.get()`.
+   */
+  prefetch: string[];
+
+  /**
+   * Isolation level for the transaction. Defaults to SerializableSnapshot.
+   */
+  isolation: Isolation;
+}
+
+/**
+ * Converts the key/value pair to a partial response that contains it. The
+ * response *will not* contain header or revision information.
+ */
+function keyValueToResponse(key: string | Buffer, value?: Buffer): RPC.IRangeResponse {
+  key = toBuffer(key);
+
+  if (!value) {
+    return <any>{ kvs: [], more: false, count: '0' };
```

Comment:
`any` usually implies some level of hackyness, what's up here?

Reply:
There is a little hackiness yes. Normally etcd would return another `header` property that contains info like the cluster revision, the ID of the server that is responding to the request, and so on. But since this is used for essentially "mocking" responses from queued writes/deletions, that information isn't available. This was the simplest way I saw to solve the problem.